### PR TITLE
Minimal hosting

### PR DESCRIPTION
**Pull request by halter73**
## Description

We're looking to collect feedback on our new "minimal hosting" API which we will present at BUILD. This adds quite a few types like `WebApplication` and `WebApplicationBuilder` as minimal alternative to `Host` and `WebHost`.

We are also planning to update the `dotnet new empty`/`dotnet new web` template for preview4 to use the new `WebApplication` type.

For this preview only, this does add temporarily a new `Configuration : IConfigurationRoot, IConfigurationBuilder` type to the `Microsoft.AspNetCore.Builder` namespace. I plan to submit a proper API proposal for to introduce this type to the `Microsoft.Extensions.Configuration` shortly.

## Customer Impact

This is entirely new API, so the only impact is giving customers an early preview version of this new minimal hosting API to try out. It does not change any of the existing Hosts so there's no impact to existing apps.

## Regression?
- [ ] Yes
- [x] No

## Risk
- [ ] High
- [ ] Medium
- [x] Low

This change merely adds new API to an existing `Microsoft.AspNetCore` assembly.

## Verification
- [x] Manual (required)
- [x] Automated

## Packaging changes reviewed?
- [ ] Yes
- [ ] No
- [x] N/A

```C#
namespace Microsoft.AspNetCore.Builder
{
    public sealed class WebApplicationBuilder
    {
        public IWebHostEnvironment Environment { get; }
        public IServiceCollection Services { get; }
        public Configuration Configuration { get; }
        public ILoggingBuilder Logging { get; }

        // Ability to configure existing web host and host
        public ConfigureWebHostBuilder WebHost { get; }
        public ConfigureHostBuilder Host { get; }

        public WebApplication Build();
    }

    public sealed class Configuration : IConfigurationRoot, IConfigurationBuilder { }

    // The .Build() methods are explicitly implemented interface method that throw NotSupportedExceptions
    public sealed class ConfigureHostBuilder : IHostBuilder { }
    public sealed class ConfigureWebHostBuilder : IWebHostBuilder { }

    public sealed class WebApplication : IHost, IApplicationBuilder, IEndpointRouteBuilder, IAsyncDisposable
    {
        // Top level properties to access common services
        public ILogger Logger { get; }
        public ICollection<string> Urls { get; }
        public IHostApplicationLifetime Lifetime { get; }
        public IServiceProvider Services { get; }
        public IConfiguration Configuration { get; }
        public IWebHostEnvironment Environment { get; }

        // Factory methods
        public static WebApplication Create(string[]? args = null);
        public static WebApplicationBuilder CreateBuilder(string[]? args = null);

        // Methods used to start the host
        public void Run(string? url = nul);
        public Task RunAsync(string? url = null);
        public Task StartAsync(CancellationToken cancellationToken = default);
        public Task StopAsync(CancellationToken cancellationToken = default);

        public ValueTask DisposeAsync();
    }
}
```

Addresses #30354

See https://github.com/featherhttp/framework to see the original versions of all of the newly added types.

The new `Configuration` type should be in Microsoft.Extensions as noted in the issue. I plan to open an API proposal to do this tomorrow. I think we'll leave it in the ASP.NET Core shared runtime temporarily for preview4.

### CODE REVIEW

File: src/DefaultBuilder/src/WebApplication.cs
Comment (halter73) [inline src/DefaultBuilder/src/WebApplication.cs:21]: What do we think? `Microsoft.AspNetCore` or `Microsoft.AspNetCore.Builder`?

Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplication.cs:21]: Everything else is in this namespace, I don't think we have anything in the root namespace

Comment (halter73) [reply src/DefaultBuilder/src/WebApplication.cs:21]: As I mention in the comment, `WebHost` is in the root namespace. After taking a closer look, I see that's [literally it.](https://docs.microsoft.com/en-us/dotnet/api/Microsoft.AspNetCore?view=aspnetcore-5.0) I agree that Builder is better in order to reduce `usings`.

File: src/DefaultBuilder/samples/SampleApp/Program.cs
Comment (MCCshreyas) [inline src/DefaultBuilder/samples/SampleApp/Program.cs:22]: Is it necessary to cast to `Func<string>`? Need some syntactic sugar here I guess because the cast looks weird! 😅

Comment (halter73) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: It's necessary for now, but we plan to fix it for C# 10. https://github.com/dotnet/csharplang/blob/main/proposals/lambda-improvements.md#implicit-conversion-to-systemdelegate

Comment (MCCshreyas) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: Yes, I thought the same earlier! 😇

File: src/DefaultBuilder/src/WebApplication.cs
Comment (Tratcher) [inline src/DefaultBuilder/src/WebApplication.cs:26]: And we've come full circle:
https://github.com/aspnet/AspNetKatana/blob/dev/src/Microsoft.Owin.Hosting/WebApp.cs

File: src/DefaultBuilder/src/WebApplication.cs
Comment (Tratcher) [inline src/DefaultBuilder/src/WebApplication.cs:109]: Many of these APIs need docs and exceptions that clarify required call ordering. E.g. "This must be called before StartAsync". Most of them should also throw if the host has been disposed.

File: src/DefaultBuilder/src/WebApplicationBuilder.cs
Comment (Tratcher) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:338]: OrdinalIgnoreCase compare all of these?

File: src/DefaultBuilder/samples/SampleApp/Program.cs
Comment (Tratcher) [inline src/DefaultBuilder/samples/SampleApp/Program.cs:22]: Did we figure out how to fix tooling like EF migrations that rely on the CreateHostBuilder pattern?

Comment (davidfowl) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: No that's out of scope for preview4

Comment (Tratcher) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: Is it tracked anywhere?

Comment (davidfowl) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: @shirhatti ?

Comment (shirhatti) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:22]: https://github.com/dotnet/aspnetcore/issues/31861

File: src/DefaultBuilder/src/WebApplication.cs
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplication.cs:63]: This should be removed.

File: src/Http/samples/MinimalSample/Program.cs
Comment (davidfowl) [inline src/Http/samples/MinimalSample/Program.cs:7]: Do you even need [FromBody] anymore?

File: src/DefaultBuilder/src/WebApplication.cs
Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplication.cs:26]: This is better though

File: src/DefaultBuilder/src/WebApplicationBuilder.cs
Comment (davidfowl) [inline src/DefaultBuilder/src/WebApplicationBuilder.cs:386]: NotFound? I'm not sure what we do normally here but we should check.

File: src/DefaultBuilder/src/WebApplication.cs
Comment (halter73) [reply src/DefaultBuilder/src/WebApplication.cs]: Fortunately, this API is being removed. You you want to specify `params string[] urls` now, you have to call an overload of `RunAsync()`.

**Comment by halter73** [comment]
@Pilchie This is up for API review Monday, but we really want this in preview4. Should we try to merge this today (pending approvals ofc), or wait for API review?

File: src/DefaultBuilder/samples/SampleApp/Program.cs
Comment (jkotalik) [inline src/DefaultBuilder/samples/SampleApp/Program.cs:20]: Idk if this is too late, but why is this Create instead of CreateAsync?

Comment (davidfowl) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:20]: This doesn't do anything asynchronous.

Comment (jkotalik) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:20]: Why await then?

Comment (davidfowl) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:20]: Async disposable.

Comment (jkotalik) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:20]: man... I get it but I don't like it.

**Comment by Pilchie** [comment]
We can retarget it to the Preview 4 branch and still merge on Monday after API review.

File: src/DefaultBuilder/src/WebApplicationBuilder.cs
Comment (halter73) [reply src/DefaultBuilder/src/WebApplicationBuilder.cs]: It looks like [we use string.Empty](https://github.com/dotnet/aspnetcore/blob/a4a3668c455fe87f731803e4d0b3abfbc4f22198/src/Hosting/Hosting/src/GenericHost/GenericWebHostBuilder.cs#L125).

There being no name is an edge case. I'm not even sure how to force it. Do we really prefer `""` over `"NotFound"`?

File: src/DefaultBuilder/samples/SampleApp/Program.cs
Comment (halter73) [reply src/DefaultBuilder/samples/SampleApp/Program.cs:20]: It's not Create that is async. `WebApplication` is `IAsyncDisposable`.

File: src/DefaultBuilder/src/WebApplicationBuilder.cs
Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplicationBuilder.cs]: I would prefer empty yes. I think when you do single file maybe?

Comment (halter73) [reply src/DefaultBuilder/src/WebApplicationBuilder.cs]: I tried single file publish on Windows to see. No difference. I bet you could force it with IL Emit or something, but that's too much work. I switched it to `string.Empty` for consistency.

File: src/DefaultBuilder/src/WebHostEnvironment.cs
Comment (BrennanConroy) [inline src/DefaultBuilder/src/WebHostEnvironment.cs:23]: Is this supposed to always be Production?

Comment (BrennanConroy) [reply src/DefaultBuilder/src/WebHostEnvironment.cs:23]: I guess that's the wrong question.

Should host config override webhost config? If you run the DefaultBuilder.SampleApp you can see it has ASPNETCORE_ENVIRONMENT set to Development but gets set to Production at https://github.com/dotnet/aspnetcore/pull/31825/files#diff-f0e6f715f99a336de90bdbeef327cb71b7ee811d00d7c1722049557adaa009c5R168

Comment (halter73) [reply src/DefaultBuilder/src/WebHostEnvironment.cs:23]: Initially, that's the default. It will be [overridden by environment variables or command line arguments by ConfigureDefaults()](https://github.com/dotnet/runtime/blob/main/src/libraries/Microsoft.Extensions.Hosting/src/HostingHostBuilderExtensions.cs#L164-L171). The flow here is pretty wild.

After the WebHostEnvironment is constructed in the WebApplicationBuilder constructor, we pass it to the ConfigureHostBuilder constructor that [calls ConfigureDefaults() on itself](https://github.com/dotnet/aspnetcore/blob/eb4ef57600c6e345fb911e341c8458ffd05c41ff/src/DefaultBuilder/src/ConfigureHostBuilder.cs#L37). ConfigureDefaults() calls ConfigureHostBuilder.ConfigureHostConfiguration which [sets the EnvironmentName as appropriate](https://github.com/dotnet/aspnetcore/blob/eb4ef57600c6e345fb911e341c8458ffd05c41ff/src/DefaultBuilder/src/ConfigureHostBuilder.cs#L69).

I'm thinking of ways to simplify the flow while keeping the shared logic in ConfigureDefaults().

Comment (halter73) [reply src/DefaultBuilder/src/WebHostEnvironment.cs:23]: I didn't see your first comment when I first replied. `DOTNET_` environment variables worked because we were bootstrapping  `ConfigureDefaults` but `ASPNETCORE_` environment variables did not work because we were not bootstrapping `ConfigureWebHostDefaults`. This is fixed by https://github.com/dotnet/aspnetcore/pull/31825/commits/b2ab79d0e930a1f03955f233aea901d911775894.

File: src/DefaultBuilder/src/WebApplication.cs
Comment (pranavkm) [inline src/DefaultBuilder/src/WebApplication.cs:67]:
```suggestion
        public ICollection<string> Addresses => ServerFeatures.Get<IServerAddressesFeature>()?.Addresses ?? throw new InvalidOperationException("Server Addresses feature could not be found.");
```

Btw, we should super consider having a `GetRequiredFeature` and be done with this issue

Comment (davidfowl) [reply src/DefaultBuilder/src/WebApplication.cs:67]: @Tratcher look your API 😄

**Comment by BrennanConroy** [comment]
FYI, working to unblock the code check with https://github.com/dotnet/aspnetcore/pull/31982

**Review by davidfowl — APPROVED** [review]

**Review by LadyNaggaga — COMMENTED** [review]
LGTM -  Let's move forward so we can get in the hands of the customers for feedback.

**Pull request merged**
